helix-core/src/test/java/org/apache/helix/task/TestDynamicTaskLoading.java

```diff
@@ -0,0 +1,123 @@
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+
+public class TestDynamicTaskLoading extends ZkTestBase {
+  HelixManager _manager;
+  MockParticipantManager _participant;
+  ClusterControllerManager _controller;
+  String _clusterName;
+
+  @BeforeClass
+  public void beforeClass() throws Exception {
+    super.beforeClass();
+    _clusterName = CLUSTER_PREFIX + "_" + getShortClassName();
+    _gSetupTool.addCluster(_clusterName, true);
+
+    _manager = HelixManagerFactory.getZKHelixManager(_clusterName, "Admin",
+        InstanceType.ADMINISTRATOR, ZK_ADDR);
+    _manager.connect();
+
+    String instanceName = "localhost_12913";
+    _gSetupTool.addInstanceToCluster(_clusterName, instanceName);
+    _participant = new MockParticipantManager(ZK_ADDR, _clusterName, instanceName);
+    StateModelFactory<StateModel> stateModelFactory = new MasterSlaveStateModelFactory(instanceName, 0);
+    StateMachineEngine stateMach = _participant.getStateMachineEngine();
+    stateMach.registerStateModelFactory("MasterSlave", stateModelFactory);
+    Map<String, TaskFactory> taskFactoryReg = new HashMap<>();
+    _participant.getStateMachineEngine().registerStateModelFactory(TaskConstants.STATE_MODEL_NAME,
+        new TaskStateModelFactory(_participant, taskFactoryReg));
+    _participant.syncStart();
+
+    _controller = new ClusterControllerManager(ZK_ADDR, _clusterName, null);
+    _controller.syncStart();
+
+    // Add task definition information as a ZNRecord.
+    ZNRecord configZnRecord = new ZNRecord(_participant.getInstanceName());
+    configZnRecord.setSimpleField("JAR_FILE", "src/test/resources/Reindex.jar");
+    configZnRecord.setSimpleField("VERSION", "1.0.0");
+    List<String> taskClasses = new ArrayList<String>();
+    taskClasses.add("com.mycompany.mocktask.MockTask");
+    configZnRecord.setListField("TASK_CLASSES", taskClasses);
+    configZnRecord.setSimpleField("TASKFACTORY", "com.mycompany.mocktask.MockTaskFactory");
+    String path = String.format("/%s/%s", _clusterName, "TASK_DEFINITION");
+    _participant.getZkClient().create(path, configZnRecord, CreateMode.PERSISTENT);
+  }
+
+  @Test
+  public void testDynamicTaskLoading() throws Exception {
+    String workflowName = TestHelper.getTestMethodName();
+    TaskDriver driver = new TaskDriver(_manager);
+    JobConfig.Builder job = new JobConfig.Builder();
+    job.setJobCommandConfigMap(Collections.singletonMap(MockTask.JOB_DELAY, "100"));
+    Workflow.Builder workflow = new Workflow.Builder(workflowName);
+    job.setWorkflow(workflowName);
+    TaskConfig taskConfig =
+        new TaskConfig(MockTask.TASK_COMMAND, new HashMap<String, String>(), null, null);
+    job.addTaskConfigMap(Collections.singletonMap(taskConfig.getId(), taskConfig));
+    job.setJobId(TaskUtil.getNamespacedJobName(workflowName, "JOB"));
+    workflow.addJob("JOB", job);
+    driver.start(workflow.build());
+
+    Thread.sleep(2000);
```

Comment:
Thanks for the PR. Good job ;-)
Is there any way to make sure the functionality of this PR? Do we need to have some sort of assert/statement to make sure it works?

Reply:
Thank you, Ali. Great point, will fix it in the next commit.